### ISSUE TITLE
gcp-cleaner throwing error GOOGLE_MANIFEST_DANGLING_PARENT_IMAGE

### ISSUE THREAD

**Issue by NiranjanHD** [GoogleCloudPlatform/gcr-cleaner#130]
### TL;DR

This is regarding the gcr-cleanup job created in the cloud-run with the container image URL: **gcr.io/gcr-cleaner/gcr-cleaner:latest**  . untagged images are not getting cleaned up but rather getting error **GOOGLE_MANIFEST_DANGLING_PARENT_IMAGE** , kindly look into this and let us know if you find more information.

### Expected behavior

gcr-cleanup job should delete untagged image without any exceptions.

### Observed behavior

Any repositories or respective subfolder listed untagged images are not getting deleted

### Debug log output

```text
failed to clean repo "<repo-path>": failed to delete "<repo-path>"@sha256:<sha-value>: DELETE https://"<repo-path>"/manifests/sha256:<sha-value>: GOOGLE_MANIFEST_DANGLING_PARENT_IMAGE: Manifest sha256:<sha-value> is referenced by image sha256:<sha-value>
```

### Additional information

Cloud logs details:

{"@type":"type.googleapis.com/google.cloud.scheduler.logging.AttemptFinished", "jobName":"<job details>", "status":"INVALID_ARGUMENT", "targetType":"HTTP", "url":"https://<gcr-path>/http"}

**Comment by ingwarsw** [comment]
Yup.. still happening..
#124 was fixing it but #125 didnt (as it promised it will)

**Comment by sethvargo** [comment]
This always only happens for multi-platform ("fat manifest") images. #125 is the best thing we can do besides giving an option to ignore this particular error. If we ignore the error, those images will remain without notice to the user.

Unfortunately there's no way to map the inter-image dependencies without downloading _all_ the manifests, which is extremely expensive. Additionally since these manifests aren't actually "images", it's unclear whether they should be removed by gcr-cleaner; they are just manifests.

I can add an option to ignore this particular error, but there's no way to avoid it short of downloading GB+ of manifests locally to build and walk a DAG. And even in that case, it's unclear how we should handle things like "keep" and "grace" in those situations.

**Comment by NiranjanHD** [comment]
thank you for sharing all the details, we will look forward to new updates if any in future.

**Comment by lauraseidler** [comment]
From my POV - tho it might not be the case for everyone - this error could be silently ignored if it still happens after a retry. I see two use cases (though please correct me if you see more):

1. Multi-platform images: The untagged manifests are parts of the parent image which is referencing them - if the parent image is not removed during the run (which should be covered by a retry), the referenced manifests shouldn't be either - so the error should be ignored if it still happens on a retry, and the manifests should remain. If the parent manifest is removed during the run, the child manifests will be removed during the retry (or at the latest, in the next run, provided the cleaner is run regularly).
2. Other kinds of manifests that reference others (e.g. via [buildx imagetools create](https://docs.docker.com/engine/reference/commandline/buildx_imagetools_create/)): I think the same logic applies - also here if the parent image is not removed during the run, the referenced manifests shouldn't be either - so the error can be ignored if it still happens on a retry, if the parent manifest is removed, retry will cover it.

I don't see any use case for these where this should throw an error and fail the run.

**Comment by eugenegoncharuk** [comment]
@lauraseidler Maybe it's worth for child images have the same tags as for parent, then current logic should work, as parent image goes before child images and will be removed first during the cleanup trigger.

**Comment by lauraseidler** [comment]
@eugenegoncharuk not sure I understand what you mean - these images are in the same repository, so they can't have the same tag. 

Also, while there's technically multiple "images", for multi arch images it's logically just one - it's just that a multi arch build produces one parent image (which is tagged) and one child image per arch (untagged), but the child images will not be used directly, they're only referenced in the parent.

**Comment by ingwarsw** [comment]
I still think that #124 was simplest and best way of dealing with it..

For sure its not fixing issue itself definitely.. but at least it fixes it in some manner ..

Maybe adding some kind of `WARN` there would be cool..

**Comment by ingwarsw** [comment]
Actually I did more tests on multi platform images, and cleaner works fine with them.

<img width="1268" alt="image" src="https://user-images.githubusercontent.com/5390156/217493628-12484878-ff8f-4eae-be8a-efa9e25fdcf5.png">

As can be seen on image ☝🏼 they dont have `creation` date but have `upload` date and thats what cleaner is using for them.

So long story short, #124 will work like a charm because:
- Cleaner cannot break manifest by deleting underlaying images because GCR will not allow it, and thats great
- If manifest will be removed (tested, both `tags` and `date` works for manifests), then assuming that you have a job thats running often and will catch that dangling images they will be deleted also

If we want to be on safe side we can change that [debug](https://github.com/GoogleCloudPlatform/gcr-cleaner/pull/124/files#diff-74ff34e86e3baf7191a3509e9a58feddf69dea7ae0c6676501539d9ce685504eR194) into info/warn, but Im not really sure its needed..

**Comment by ingwarsw** [comment]
After much more testing and reading code I have found out Im still using pre last changes version.
And new one (`0.11.1`) actually skips that `GOOGLE_MANIFEST_DANGLING_PARENT_IMAGE` errors.
And it works like a charm.
